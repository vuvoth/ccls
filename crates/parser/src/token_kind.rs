use logos::Logos;
use serde::Serialize;

#[derive(Logos, Debug, PartialEq, Clone, Copy, Eq, PartialOrd, Ord, Hash, Serialize)]
#[allow(non_camel_case_types)]
#[repr(u16)]
pub enum TokenKind {
    // Error
    #[error]
    Error = 0,
    // Comments
    #[regex(r"//[^\r\n]*")]
    CommentLine,
    #[token("/*")]
    CommentBlockOpen,
    #[token("*/")]
    CommentBlockClose,
    // Trivial
    #[regex("[ \t]+")]
    WhiteSpace,
    #[regex(r"\r?\n")]
    EndLine,
    // Pragma
    Pragma,
    #[token("pragma")]
    PragmaKw,
    #[token("circom")]
    Circom,
    #[regex("2.[0-9].[0-9]")]
    Version,
    // Literals
    #[regex("[0-9]+")]
    Number,
    #[regex("[$_]*[a-zA-Z][a-zA-Z0-9_$]*")]
    Identifier,
    #[regex(r#""[^"]*""#)]
    CircomString,
    // Brackets
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("{")]
    LCurly,
    #[token("}")]
    RCurly,
    #[token("[")]
    LBracket,
    #[token("]")]
    RBracket,
    // Punctuation
    #[token(";")]
    Semicolon,
    #[token(",")]
    Comma,
    #[token(".")]
    Dot,
    // Boolean operators
    #[token("&&")]
    BoolAnd,
    #[token("||")]
    BoolOr,
    #[token("!")]
    Not,
    // Relational operators
    #[token("==")]
    Equal,
    #[token("!=")]
    NotEqual,
    #[token("<")]
    LessThan,
    #[token(">")]
    GreaterThan,
    #[token("<=")]
    LessThanAndEqual,
    #[token(">=")]
    GreaterThanAndEqual,
    // Arithmetic operators
    #[token("+")]
    Add,
    #[token("-")]
    Sub,
    #[token("*")]
    Mul,
    #[token("**")]
    Power,
    #[token("/")]
    Div,
    #[token("\\")]
    IntDiv,
    #[token("%")]
    Mod,
    // Combined arithmetic assignment
    #[token("+=")]
    AddAssign,
    #[token("-=")]
    SubAssign,
    #[token("*=")]
    MulAssign,
    #[token("**=")]
    PowerAssign,
    #[token("/=")]
    DivAssign,
    #[token(r"\=")]
    IntDivAssign,
    #[token("%=")]
    ModAssign,
    #[token("++")]
    UnitInc,
    #[token("--")]
    UnitDec,
    // Bitwise operators
    #[token("&")]
    BitAnd,
    #[token("|")]
    BitOr,
    #[token("~")]
    BitNot,
    #[token("^")]
    BitXor,
    #[token(">>")]
    ShiftR,
    #[token("<<")]
    ShiftL,
    // Combined bitwise assignments
    #[token("&=")]
    BitAndAssign,
    #[token("|=")]
    BitOrAssign,
    #[token("~=")]
    BitNotAssign,
    #[token("^=")]
    BitXorAssign,
    #[token(">>=")]
    ShiftRAssign,
    #[token("<<=")]
    ShiftLAssign,
    // Assign
    #[token("=")]
    Assign,
    #[token("===")]
    EqualSignal,
    #[token("-->")]
    LAssignSignal,
    #[token("==>")]
    LAssignContraintSignal,
    #[token("<--")]
    RAssignSignal,
    #[token("<==")]
    RAssignConstraintSignal,
    // Conditional expressions
    #[token("?")]
    MarkQuestion,
    #[token(":")]
    Colon,
    // Keywords
    #[token("template")]
    TemplateKw,
    #[token("function")]
    FunctionKw,
    #[token("component")]
    ComponentKw,
    #[token("main")]
    MainKw,
    #[token("public")]
    PublicKw,
    #[token("signal")]
    SignalKw,
    #[token("var")]
    VarKw,
    #[token("include")]
    IncludeKw,
    #[token("input")]
    InputKw,
    #[token("output")]
    OutputKw,
    #[token("log")]
    LogKw,
    // Statement keywords
    #[token("if")]
    IfKw,
    #[token("else")]
    ElseKw,
    #[token("for")]
    ForKw,
    #[token("while")]
    WhileKw,
    #[token("return")]
    ReturnKw,
    #[token("assert")]
    AssertKw,
    // Statements
    IfStatement,
    AssertStatement,
    LogStatement,
    ReturnStatement,
    AssignStatement,
    ForLoop,
    WhileLoop,
    // Program
    CircomProgram,
    // Function
    FunctionDef,
    FunctionName,
    // Template
    TemplateDef,
    TemplateName,
    // Signal
    SignalDecl,
    InputSignalDecl,
    OutputSignalDecl,
    SignalHeader,
    SignalIdentifier,
    // Variable
    VarDecl,
    VarIdentifier,
    // Component
    ComponentDecl,
    ComponentCall,
    ComponentIdentifier,
    SignalOfComponent,
    // Expression
    ExpressionAtom,
    Expression,
    // Complex token kind
    Block,
    ParameterList,
    Call,
    TenaryConditional,
    Condition,
    Statement,
    StatementList,
    ArrayQuery,
    ParserError,
    BlockComment,
    EOF,
    ROOT,
    __LAST,
}

impl From<u16> for TokenKind {
    #[inline]
    fn from(d: u16) -> TokenKind {
        assert!(d <= (TokenKind::__LAST as u16));
        unsafe { std::mem::transmute::<u16, TokenKind>(d) }
    }
}

impl From<rowan::SyntaxKind> for TokenKind {
    fn from(value: rowan::SyntaxKind) -> Self {
        match value {
            rowan::SyntaxKind(id) => TokenKind::from(id),
        }
    }
}

impl From<TokenKind> for u16 {
    #[inline]
    fn from(k: TokenKind) -> u16 {
        k as u16
    }
}

impl From<TokenKind> for rowan::SyntaxKind {
    fn from(kind: TokenKind) -> Self {
        Self(kind as u16)
    }
}

impl TokenKind {
    // a + 10 --> a and 10 are literals
    pub fn is_literal(self) -> bool {
        matches!(self, Self::Number | Self::Identifier)
    }

    // these tokens have the lowest priority
    // <identifier1> infix_operator <identifier2>
    // eg: a + b --> + is an infix token
    pub fn infix(self) -> Option<(u16, u16)> {
        match self {
            // arithmetic operators
            Self::Power => Some((99, 100)),
            Self::Mul | Self::Div | Self::IntDiv | Self::Mod => Some((94, 95)),
            Self::Add | Self::Sub => Some((89, 90)),
            // shift bitwise operators
            Self::ShiftL | Self::ShiftR => Some((84, 85)),
            // relational operators
            Self::LessThan
            | Self::GreaterThan
            | Self::LessThanAndEqual
            | Self::GreaterThanAndEqual => Some((79, 80)),
            Self::Equal | Self::NotEqual => Some((74, 75)),
            // other bitwise operators
            Self::BitAnd => Some((69, 70)),
            Self::BitXor => Some((64, 65)), // exclusive or
            Self::BitOr => Some((59, 60)),
            // boolean operators
            Self::BoolAnd => Some((54, 55)),
            Self::BoolOr => Some((49, 50)),
            // ----------
            // TODO: how about conditional operation ( ? : )
            // associativity: right to left [ a ? b : c --> ??? ]

            // ----------
            // associativity: right to left [ a = b = c --> a = (b = c) ]
            // DO NOT CONSIDER ASSIGMENT OPERATORS AS INFIX TOKENS
            /*
            // assignment operators
            Self::Assign
            // signal assigment operators
            | Self::EqualSignal
            | Self::LAssignSignal
            | Self::LAssignContraintSignal
            | Self::RAssignSignal
            | Self::RAssignConstraintSignal
            // bitwise asignment operators
            | Self::BitOrAssign
            | Self::BitXorAssign
            | Self::BitAndAssign
            | Self::ShiftLAssign
            | Self::ShiftRAssign
            // arithmetic asignament operators
            | Self::AddAssign
            | Self::SubAssign
            | Self::MulAssign
            | Self::DivAssign
            | Self::IntDivAssign
            | Self::ModAssign
            | Self::PowerAssign => Some((44, 45)),
            */
            // TODO: how about comma (expression separator)
            Self::Comma => Some((39, 40)),
            // not an infix operator
            _ => None,
        }
    }

    // priority: post > pre > in
    // associativity: right to left [ --!a --> --(!a) ]
    // prefix_operator <literal>
    // eg: -10, !a, ++a, --a
    pub fn prefix(self) -> Option<u16> {
        match self {
            Self::UnitDec | Self::UnitInc | Self::Sub | Self::Add | Self::Not | Self::BitNot => {
                Some(200)
            }

            _ => None,
        }
    }

    // these tokens have the highest priority
    // <literal> postfix_operator
    // eg: a[10], b++, c.att1
    pub fn postfix(self) -> Option<u16> {
        match self {
            Self::LParen // function call
            | Self::LBracket // array subscript
            | Self::Dot // attribute access
            | Self::UnitDec | Self::UnitInc => Some(300),

            _ => None,
        }
    }

    pub fn is_declaration_kw(self) -> bool {
        matches!(self, Self::VarKw | Self::ComponentKw | Self::SignalKw)
    }

    pub fn is_assign_token(self) -> bool {
        matches!(
            self,
            Self::Assign
            // signal assigment operators
            | Self::EqualSignal
            | Self::LAssignSignal
            | Self::LAssignContraintSignal
            | Self::RAssignSignal
            | Self::RAssignConstraintSignal
            // bitwise asignment operators
            | Self::BitOrAssign
            | Self::BitXorAssign
            | Self::BitAndAssign
            | Self::ShiftLAssign
            | Self::ShiftRAssign
            // arithmetic asignament operators
            | Self::AddAssign
            | Self::SubAssign
            | Self::MulAssign
            | Self::DivAssign
            | Self::IntDivAssign
            | Self::ModAssign
            | Self::PowerAssign // unit inc/dec
                                // | Self::UnitInc
                                // | Self::UnitDec
        )
    }

    pub fn is_inline_assign_signal(self) -> bool {
        matches!(
            self,
            Self::Assign | Self::RAssignSignal | Self::RAssignConstraintSignal
        )
    }

    pub fn is_var_assign(self) -> bool {
        matches!(self, Self::Assign)
    }

    pub fn is_trivial(self) -> bool {
        matches!(
            self,
            Self::WhiteSpace | Self::EndLine | Self::CommentLine | Self::BlockComment | Self::Error
        )
    }
}
