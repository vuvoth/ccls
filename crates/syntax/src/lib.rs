pub mod abstract_syntax_tree;
pub mod syntax;
pub mod syntax_node;
mod view_syntax;
mod utils;
