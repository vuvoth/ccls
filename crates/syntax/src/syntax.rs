use parser::grammar::entry::Scope;
use parser::input::Input;
use parser::output::{Child, Output};
use parser::parser::Parser;
use parser::token_kind::TokenKind;
use rowan::{GreenNode, GreenNodeBuilder};

pub use rowan::{
    api::Preorder, Direction, NodeOrToken, SyntaxText, TextRange, TextSize, TokenAtOffset,
    WalkEvent,
};

use crate::syntax_node::SyntaxNode;

pub struct SyntaxTreeBuilder<'a> {
    builder: GreenNodeBuilder<'static>,
    input: &'a Input<'a>,
}

impl<'a> SyntaxTreeBuilder<'a> {
    pub fn new(input: &'a Input) -> Self {
        Self {
            builder: GreenNodeBuilder::new(),
            input,
        }
    }
    pub fn build_rec(&mut self, tree: &Output) {
        self.builder.start_node(tree.kind().into());
        for child in tree.children() {
            match child {
                Child::Token(token_id) => {
                    let token_kind = self.input.kind_of(*token_id);
                    // TODO: return Error to replace .unwrap()
                    let token_value = self.input.token_value(*token_id).unwrap();
                    self.builder.start_node(token_kind.into());
                    self.builder.token(token_kind.into(), token_value);
                    self.builder.finish_node();
                }
                Child::Tree(child_tree) => self.build_rec(child_tree),
                Child::Error(error) => {
                    let token_kind = TokenKind::Error;
                    let token_value = error.as_str();

                    self.builder.start_node(token_kind.into());
                    self.builder.token(token_kind.into(), token_value);
                    self.builder.finish_node();
                }
            }
        }

        self.builder.finish_node();
    }

    pub fn build(&mut self, tree: Output) {
        self.build_rec(&tree);
    }

    pub fn finish(self) -> GreenNode {
        self.builder.finish()
    }

    pub fn syntax_tree(source: &str) -> SyntaxNode {
        let input = Input::new(source);

        let output = Parser::parsing(&input);

        let mut builder = SyntaxTreeBuilder::new(&input);
        builder.build(output);
        let green = builder.finish();
        SyntaxNode::new_root(green)
    }
}

pub fn syntax_node_from_source(source: &str, scope: Scope) -> SyntaxNode {
    let input = Input::new(&source);
    let output = Parser::parsing_with_scope(&input, scope);

    // output is a tree whose node is index of token, no content of token
    // convert output into green node
    let mut builder = SyntaxTreeBuilder::new(&input);
    builder.build(output);
    let green = builder.finish();

    // then cast green node into syntax node
    let syntax = SyntaxNode::new_root(green);

    syntax
}

#[cfg(test)]
mod tests {
    use crate::test_syntax;
    use parser::grammar::entry::Scope;

    #[test]
    fn pragma_happy_test() {
        test_syntax!("/src/test_files/happy/pragma.circom", Scope::Pragma);
    }

    #[test]
    fn template_happy_test() {
        // SOURCE & EXPECTED RESULT
        test_syntax!("/src/test_files/happy/template.circom", Scope::Template);
    }

    #[test]
    fn block_happy_test() {
        test_syntax!("/src/test_files/happy/block.circom", Scope::Block);
    }

    #[test]
    fn comment_happy_test() {
        test_syntax!(
            "/src/test_files/happy/block_comment.circom",
            Scope::CircomProgram
        );
        test_syntax!(
            "/src/test_files/happy/line_comment.circom",
            Scope::CircomProgram
        );
    }

    #[test]
    fn full_circom_program() {
        test_syntax!(
            "/src/test_files/happy/full_circom_program.circom",
            Scope::CircomProgram
        );
    }
}
