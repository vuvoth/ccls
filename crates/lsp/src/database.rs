use std::{
    collections::HashMap,
    hash::{Hash, Hasher},
    path::PathBuf,
};

use std::collections::hash_map::DefaultHasher;

use lsp_types::{Position, Range, Url};

use rowan::{ast::AstNode, TextSize};
use syntax::{
    abstract_syntax_tree::{
        AstCircomProgram, AstComponentDecl, AstInputSignalDecl, AstOutputSignalDecl, AstSignalDecl,
        AstTemplateDef, AstVarDecl,
    },
    syntax_node::{SyntaxNode, SyntaxToken},
};

/**
* We will store
* Open data -> Parse -> output -> Syntax -> analyzer -> db{
   FileID {
       Template {
           signal,

       }
   }

                               value
   Template map: { Hash(FileID, token) -> Template}
   Vars map: {Hash(FileID, template, token)} -> Var}
   Component map {Hash(FileID, template, token)} -> ComponentInfo
   Signals map {Hash(FileID, template, token)} -> Signal



}
*/

#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
pub struct FileId(pub u64);

#[derive(Clone)]
pub struct FileDB {
    pub file_id: FileId,
    pub file_path: Url,
    pub end_line_vec: Vec<u32>,
}

use path_absolutize::*;

impl FileDB {
    pub fn create(content: &str, file_path: Url) -> Self {
        let mut hasher = DefaultHasher::new();
        file_path
            .to_file_path()
            .unwrap()
            .absolutize()
            .unwrap()
            .hash(&mut hasher);
        Self::new(FileId(hasher.finish()), content, file_path)
    }

    pub(super) fn new(file_id: FileId, content: &str, file_path: Url) -> Self {
        let mut file_utils = Self {
            file_id,
            file_path,
            end_line_vec: Vec::new(),
        };

        for (id, c) in content.chars().enumerate() {
            if c == '\n' {
                file_utils.end_line_vec.push(id as u32);
            }
        }

        file_utils
    }

    pub fn get_path(&self) -> PathBuf {
        let p = self.file_path.path();
        PathBuf::from(p)
    }

    pub fn off_set(&self, position: Position) -> TextSize {
        if position.line == 0 {
            return position.character.into();
        }
        (self.end_line_vec[position.line as usize - 1] + position.character + 1).into()
    }

    pub fn position(&self, off_set: TextSize) -> Position {
        let line = match self.end_line_vec.binary_search(&(off_set.into())) {
            Ok(l) => l,
            Err(l) => l,
        };

        Position::new(
            line as u32,
            if line > 0 {
                (u32::from(off_set)) - self.end_line_vec[line - 1] - 1
            } else {
                off_set.into()
            },
        )
    }

    pub fn range(&self, syntax: &SyntaxNode) -> Range {
        let syntax_range = syntax.text_range();
        Range {
            start: self.position(syntax_range.start()),
            end: self.position(syntax_range.end()),
        }
    }
}

#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
pub struct Id(pub u64);

pub trait TokenId {
    fn token_id(&self) -> Id;
}

impl TokenId for SyntaxNode {
    fn token_id(&self) -> Id {
        let mut hasher = DefaultHasher::new();
        self.to_string().hash(&mut hasher);
        Id(hasher.finish())
    }
}

impl TokenId for SyntaxToken {
    fn token_id(&self) -> Id {
        let mut hasher = DefaultHasher::new();
        self.to_string().hash(&mut hasher);
        Id(hasher.finish())
    }
}

#[derive(Debug, Clone)]
pub struct SemanticLocations(pub HashMap<Id, Vec<Range>>);

impl Default for SemanticLocations {
    fn default() -> Self {
        Self::new()
    }
}

impl SemanticLocations {
    pub fn insert(&mut self, token_id: Id, range: Range) {
        if let Some(locations) = self.0.get_mut(&token_id) {
            locations.push(range);
        } else {
            self.0.insert(token_id, vec![range]);
        }
    }
    pub fn new() -> Self {
        Self(HashMap::new())
    }
}

#[derive(Debug, Clone)]
pub struct TemplateDataSemantic {
    pub signal: SemanticLocations,
    pub variable: SemanticLocations,
    pub component: SemanticLocations,
}

impl TemplateDataSemantic {
    fn new() -> Self {
        Self {
            signal: SemanticLocations::new(),
            variable: SemanticLocations::new(),
            component: SemanticLocations::new(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct SemanticData {
    pub template: SemanticLocations,
    pub template_data_semantic: HashMap<Id, TemplateDataSemantic>,
}

pub enum TemplateDataInfo {
    Signal((Id, Range)),
    Variable((Id, Range)),
    Component((Id, Range)),
}
pub enum SemanticInfo {
    Template((Id, Range)),
    TemplateData((Id, TemplateDataInfo)),
}

#[derive(Debug, Clone)]
pub struct SemanticDB {
    pub semantic: HashMap<FileId, SemanticData>,
}

impl Default for SemanticDB {
    fn default() -> Self {
        Self::new()
    }
}

impl SemanticDB {
    pub fn new() -> Self {
        Self {
            semantic: HashMap::new(),
        }
    }

    pub fn insert(&mut self, file_id: FileId, semantic_info: SemanticInfo) {
        let semantic = self.semantic.entry(file_id).or_insert(SemanticData {
            template: SemanticLocations::new(),
            template_data_semantic: HashMap::new(),
        });

        match semantic_info {
            SemanticInfo::Template((id, range)) => {
                semantic.template.insert(id, range);
            }
            SemanticInfo::TemplateData((template_id, template_data_info)) => {
                let template_semantic = semantic
                    .template_data_semantic
                    .entry(template_id)
                    .or_insert(TemplateDataSemantic::new());

                match template_data_info {
                    TemplateDataInfo::Component((id, r)) => {
                        template_semantic.component.insert(id, r)
                    }
                    TemplateDataInfo::Variable((id, r)) => template_semantic.variable.insert(id, r),
                    TemplateDataInfo::Signal((id, r)) => template_semantic.signal.insert(id, r),
                }
            }
        }
    }

    pub fn circom_program_semantic(
        &mut self,
        file_db: &FileDB,
        abstract_syntax_tree: &AstCircomProgram,
    ) {
        for template in abstract_syntax_tree.template_list() {
            if let Some(name) = template.name() {
                let template_id = name.syntax().token_id();
                self.insert(
                    file_db.file_id,
                    SemanticInfo::Template((template_id, file_db.range(template.syntax()))),
                );
                self.template_semantic(file_db, &template);
            }
        }
    }

    pub fn template_semantic(&mut self, file_db: &FileDB, ast_template: &AstTemplateDef) {
        let template_id = ast_template.syntax().token_id();

        if let Some(statements) = ast_template.statements() {
            for signal in statements.find_children::<AstInputSignalDecl>() {
                if let Some(name) = signal.name() {
                    self.insert(
                        file_db.file_id,
                        SemanticInfo::TemplateData((
                            template_id,
                            TemplateDataInfo::Signal((
                                name.syntax().token_id(),
                                file_db.range(signal.syntax()),
                            )),
                        )),
                    );
                }
            }
            for signal in statements.find_children::<AstOutputSignalDecl>() {
                if let Some(name) = signal.name() {
                    self.insert(
                        file_db.file_id,
                        SemanticInfo::TemplateData((
                            template_id,
                            TemplateDataInfo::Signal((
                                name.syntax().token_id(),
                                file_db.range(signal.syntax()),
                            )),
                        )),
                    );
                }
            }

            for signal in statements.find_children::<AstSignalDecl>() {
                if let Some(name) = signal.name() {
                    self.insert(
                        file_db.file_id,
                        SemanticInfo::TemplateData((
                            template_id,
                            TemplateDataInfo::Signal((
                                name.syntax().token_id(),
                                file_db.range(signal.syntax()),
                            )),
                        )),
                    );
                }
            }

            for var in statements.find_children::<AstVarDecl>() {
                if let Some(name) = var.name() {
                    self.insert(
                        file_db.file_id,
                        SemanticInfo::TemplateData((
                            template_id,
                            TemplateDataInfo::Variable((
                                name.syntax().token_id(),
                                file_db.range(var.syntax()),
                            )),
                        )),
                    );
                }
            }

            for component in statements.find_children::<AstComponentDecl>() {
                if let Some(component_var) = component.component_identifier() {
                    if let Some(name) = component_var.name() {
                        self.insert(
                            file_db.file_id,
                            SemanticInfo::TemplateData((
                                template_id,
                                TemplateDataInfo::Component((
                                    name.syntax().token_id(),
                                    file_db.range(component.syntax()),
                                )),
                            )),
                        );
                    }
                }
            }
        }
    }
}

impl SemanticData {
    pub fn lookup_signal(&self, template_id: Id, signal: &SyntaxToken) -> Option<&Vec<Range>> {
        if let Some(semantic_template) = self.template_data_semantic.get(&template_id) {
            return semantic_template.signal.0.get(&signal.token_id());
        }
        None
    }

    // TODO: remove duplicate code here.
    pub fn lookup_variable(&self, template_id: Id, variable: &SyntaxToken) -> Option<&Vec<Range>> {
        if let Some(semantic_template) = self.template_data_semantic.get(&template_id) {
            return semantic_template.variable.0.get(&variable.token_id());
        }
        None
    }

    pub fn lookup_component(
        &self,
        template_id: Id,
        component: &SyntaxToken,
    ) -> Option<&Vec<Range>> {
        if let Some(semantic_template) = self.template_data_semantic.get(&template_id) {
            return semantic_template.component.0.get(&component.token_id());
        }
        None
    }
}

#[cfg(test)]
mod tests {

    use std::path::Path;

    use ::syntax::{abstract_syntax_tree::AstCircomProgram, syntax::SyntaxTreeBuilder};
    use lsp_types::{Position, Url};

    use rowan::ast::AstNode;

    use crate::database::{FileDB, FileId};

    use super::TokenId;

    #[test]
    fn file_id_test() {
        let file_1 = FileDB::create("a", Url::from_file_path(Path::new("/a/../a/c")).unwrap());
        let file_2 = FileDB::create("a", Url::from_file_path(Path::new("/a/c")).unwrap());

        assert_eq!(file_1.file_id, file_2.file_id);
    }
    #[test]
    fn token_id_hash_test() {
        let source: String = r#"pragma circom 2.0.0;

        
        template Multiplier2 () {}
        template Multiplier2 () {} 
        "#
        .to_string();

        let syntax = SyntaxTreeBuilder::syntax_tree(&source);

        if let Some(ast) = AstCircomProgram::cast(syntax) {
            let templates = ast.template_list();
            let first_id = templates[0].syntax().token_id();
            let second_id = templates[1].syntax().token_id();

            assert_eq!(first_id, second_id);
        }
    }
    #[test]
    fn off_set_test() {
        let str = r#"
one
two
three
       "#;

        let file_utils = FileDB::new(
            FileId(1),
            str,
            Url::from_file_path(Path::new("/tmp.txt")).unwrap(),
        );

        let position = Position::new(0, 1);

        assert_eq!(file_utils.off_set(position), 1.into());

        let position = Position::new(1, 1);

        assert_eq!(file_utils.off_set(position), 2.into());
    }

    #[test]
    fn position_test() {
        let str = r#"
        one
        two
        three
               "#;

        // 0, 4, 8
        let file_utils = FileDB::new(
            FileId(1),
            str,
            Url::from_file_path(Path::new("/tmp.txt")).unwrap(),
        );
        assert_eq!(Position::new(1, 1), file_utils.position(2.into()));
        assert_eq!(Position::new(0, 0), file_utils.position(0.into()));
    }
}
