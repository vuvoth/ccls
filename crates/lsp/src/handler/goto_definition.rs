use lsp_types::Location;
use lsp_types::Position;
use lsp_types::Range;
use lsp_types::Url;
use parser::token_kind::TokenKind;
use rowan::ast::AstNode;
use rowan::SyntaxText;

use syntax::abstract_syntax_tree::AstComponentCall;
use syntax::abstract_syntax_tree::AstInclude;
use syntax::abstract_syntax_tree::AstTemplateDef;
use syntax::abstract_syntax_tree::AstTemplateName;
use syntax::abstract_syntax_tree::{AstCircomProgram, AstComponentDecl};
use syntax::syntax_node::SyntaxNode;
use syntax::syntax_node::SyntaxToken;

use crate::database::{FileDB, SemanticData, TokenId};

// find the first ancestor with given kind of a syntax token
pub fn lookup_node_wrap_token(ast_type: TokenKind, token: &SyntaxToken) -> Option<SyntaxNode> {
    let mut p = token.parent();
    while let Some(t) = p {
        if t.kind() == ast_type {
            return Some(t);
        }
        p = t.parent();
    }
    None
}

// return an Identifier/CircomString token at a position
pub fn lookup_token_at_postion(
    file: &FileDB,
    ast: &AstCircomProgram,
    position: Position,
) -> Option<SyntaxToken> {
    let off_set = file.off_set(position);
    ast.syntax().token_at_offset(off_set).find_map(|token| {
        let kind = token.kind();

        if kind == TokenKind::Identifier {
            return Some(token);
        }

        if kind == TokenKind::CircomString {
            return Some(token);
        }
        None
    })
}

// find all template name (in component declaration) which are used inside a template
pub fn lookup_component(template: &AstTemplateDef, text: SyntaxText) -> Option<AstTemplateName> {
    if let Some(statements) = template.statements() {
        for component in statements.find_children::<AstComponentDecl>() {
            if let Some(iden) = component.component_identifier() {
                if iden.name().unwrap().syntax().text() == text {
                    return component.template();
                }
            }
        }
    }
    None
}

// if token in an include statement
// add lib path (location of source code of that library) into result
pub fn jump_to_lib(file: &FileDB, token: &SyntaxToken) -> Vec<Location> {
    if let Some(include_lib) = lookup_node_wrap_token(TokenKind::IncludeKw, token) {
        if let Some(ast_include) = AstInclude::cast(include_lib) {
            if let Some(abs_lib_ans) = ast_include.lib() {
                let lib_path = file
                    .get_path()
                    .parent()
                    .unwrap()
                    .join(abs_lib_ans.value())
                    .clone();
                let lib_url = Url::from_file_path(lib_path.clone()).unwrap();
                return vec![Location::new(lib_url, Range::default())];
            }
        }
    }

    Vec::new()
}

pub fn lookup_definition(
    file: &FileDB,
    ast: &AstCircomProgram,
    semantic_data: &SemanticData,
    token: &SyntaxToken,
) -> Vec<Location> {
    let template_list = ast.template_list();
    // TODO: extract function list

    let mut res = Vec::new();

    if token.kind() == TokenKind::CircomString {
        return jump_to_lib(file, token);
    }

    // signal from other template
    // eg: in1, in2 from component call mul(in1, in2)
    let mut signal_outside = false;

    if let Some(component_call) = lookup_node_wrap_token(TokenKind::ComponentCall, token) {
        // find template called.
        if let Some(ast_component_call) = AstComponentCall::cast(component_call) {
            if let Some(signal) = ast_component_call.signal() {
                // if target token is the parameter of a component call
                // TODO: go to params in template!!! (failed)
                if signal.syntax().text() == token.text() {
                    signal_outside = true;
                    // lookup template of component
                    if let Some(current_template) =
                        lookup_node_wrap_token(TokenKind::TemplateDef, token)
                    {
                        if let Some(ast_template_name) = lookup_component(
                            &AstTemplateDef::cast(current_template).unwrap(),
                            ast_component_call.component_name().unwrap().syntax().text(),
                        ) {
                            if let Some(other_template) =
                                ast.get_template_by_name(&ast_template_name)
                            {
                                let template_id = other_template.syntax().token_id();
                                if let Some(semantic) =
                                    semantic_data.template_data_semantic.get(&template_id)
                                {
                                    if let Some(tmp) =
                                        semantic.signal.0.get(&signal.syntax().token_id())
                                    {
                                        res.extend(tmp)
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    if !signal_outside {
        // look up token in template information
        // (template name, signal/variable/component in template)
        for template in template_list {
            let template_name = template.name().unwrap();
            if template_name.name().unwrap().syntax().text() == token.text() {
                let range = file.range(template.syntax());
                res.push(range);
            }

            if !template
                .syntax()
                .text_range()
                .contains_range(token.text_range())
            {
                continue;
            }

            let template_id = template.syntax().token_id();

            if let Some(data) = semantic_data.lookup_signal(template_id, token) {
                res.extend(data);
            }

            if let Some(data) = semantic_data.lookup_variable(template_id, token) {
                res.extend(data);
            }

            if let Some(component_decl) = semantic_data.lookup_component(template_id, token) {
                res.extend(component_decl);
            }
        }

        // TODO: look up token in function information
        // (function name, signal/variable/component in function)
    }

    res.into_iter()
        .map(|range| Location::new(file.file_path.clone(), range))
        .collect()
}

#[cfg(test)]
mod tests {
    use std::path::Path;

    use lsp_types::Url;
    use parser::token_kind::TokenKind;
    use rowan::ast::AstNode;
    use syntax::{
        abstract_syntax_tree::{AstCircomProgram, AstInputSignalDecl},
        syntax::SyntaxTreeBuilder,
    };

    use crate::{database::FileDB, handler::goto_definition::lookup_node_wrap_token};

    use super::lookup_token_at_postion;

    fn get_source_from_path(file_path: &str) -> String {
        let crate_path = std::env::var("CARGO_MANIFEST_DIR").unwrap();
        let full_path = format!("{}{}", crate_path, file_path);
        let source = std::fs::read_to_string(&full_path).expect(&full_path);

        source
    }

    #[test]
    fn goto_decl_test() {
        let file_path = "/src/test_files/handler/templates.circom";
        let source = get_source_from_path(file_path);
        let file = FileDB::create(&source, Url::from_file_path(Path::new("/tmp")).unwrap());

        let syntax_node = SyntaxTreeBuilder::syntax_tree(&source);

        if let Some(program_ast) = AstCircomProgram::cast(syntax_node) {
            let inputs = program_ast.template_list()[0]
                .func_body()
                .unwrap()
                .statement_list()
                .unwrap()
                .find_children::<AstInputSignalDecl>();
            let signal_name = inputs[0].signal_identifier().unwrap().name().unwrap();

            let tmp = signal_name.syntax().text_range().start();

            if let Some(token) = lookup_token_at_postion(&file, &program_ast, file.position(tmp)) {
                let wrap_token = lookup_node_wrap_token(TokenKind::TemplateDef, &token);

                let string_syntax_node = match wrap_token {
                    None => "None".to_string(),
                    Some(syntax_node) => format!("{}", syntax_node),
                };

                insta::assert_snapshot!("test_lookup_node_wrap_token", string_syntax_node);
            }
        }
    }

    #[test]
    fn url_test() {
        let url = Url::from_file_path(Path::new("/hello/abc.tx"));
        let binding = url.unwrap();
        let path = binding.path();
        let parent = Path::new(path).parent().unwrap().to_str().unwrap();

        assert_eq!("/hello", parent);
    }
}
