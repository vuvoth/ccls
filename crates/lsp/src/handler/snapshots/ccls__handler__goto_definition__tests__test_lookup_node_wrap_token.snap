---
source: crates/lsp/src/handler/goto_definition.rs
expression: string_syntax_node
---
template X() {
            signal x[100];
            signal input x expect Semicolon but got Assign= expect Semicolon but got Number10;
           component x = Multiplier2();
           component y = X();
           component y = Multiplier2();
           component z = Multiplier2();
              
        }
